presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -184,6 +185,80 @@ else if (type instanceof TimestampWithTimeZoneType) {
         return Optional.of(unmodifiableList(row));
     }
 
+    private Object getLegacyValue(Object value, Type type)
+    {
+        if (value == null) {
+            return null;
+        }
+
+        if (!supportsParametricDateTime) {
+            // for legacy clients we need to round timestamp and timestamp with timezone to default precision (3)
+
+            if (type instanceof TimestampType) {
+                return ((SqlTimestamp) value).roundTo(3);
+            }
+
+            if (type instanceof TimestampWithTimeZoneType) {
+                return ((SqlTimestampWithTimeZone) value).roundTo(3);
+            }
+        }
+
+        if (type instanceof ArrayType) {
+            Type elementType = ((ArrayType) type).getElementType();
+
+            if (!(elementType instanceof TimestampType || elementType instanceof TimestampWithTimeZoneType)) {
+                return value;
+            }
+
+            List<Object> values = (List<Object>) value;
+            ImmutableList.Builder<Object> result = ImmutableList.builder();
+
+            for (int i = 0; i < values.size(); i++) {
+                result.add(getLegacyValue(values.get(i), elementType));
+            }
+
+            return result.build();
```

Comment:
This can be simplified to:

```suggestion
            return ((List<Object>) value).stream()
                    .map(element -> getLegacyValue(value, elementType))
                    .collect(toImmutableList());
```